### PR TITLE
Added SIP that reproduces the checksum algorithm

### LINKED ISSUE #123: SIP for multi-file checksum algorithm
In the future we'd like to use [Decentralised Identifiers](https://www.w3.org/TR/did-core/) to refer to Snaps.

It's going to use multi-file checksum that we write to `snap.manfiest.json`, like so `did:snap:8uciDR11gmEm8QDs/sHYPeICmkpHOE4qVe+7zF0O7Ew=`.

The checksum algorithm is currently standardised only in [`snap-utils`](https://github.com/MetaMask/snaps/blob/main/packages/snaps-utils/src/checksum.ts) and is based on [`fast-json-stable-stringify`](https://github.com/MetaMask/snaps/blob/6e0257741c7eb0fb71df5826fcfabb7658abd03e/packages/snaps-utils/src/snaps.ts#L165) npm package.

We should write a SIP that provides all steps to make such a reproducible checksum

### PR DESCRIPTION
**Pull request by ritave**
This SIP describes an algorithm used to checksum a snap in a reproducible manner in a way that includes all files required to run the snap.

fixes #123

### CODE REVIEW

File: assets/sip-124/implementation/implementation.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import stableStringify from "fast-json-stable-stringify";
+import { sha256 } from "@noble/hashes/sha256";
+import { concatBytes } from "@metamask/utils";
+import { base64 } from "@scure/base";
+
+export type VFile = { path: string; contents: string };
```
Comment (FrederikBolding) [inline assets/sip-124/implementation/implementation.ts:6]:
```suggestion
export type VFile = { path: string; contents: string | Uint8Array };
```

File: assets/sip-124/implementation/implementation.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  delete structure.result.source.shasum;
+  return { path, contents: stableStringify(structure) };
+}
+
+export function checksumFiles(manifest: VFile, auxiliary: VFile[]): string {
+  return base64.encode(
+    sha256(
+      concatBytes(
+        [stableManifest(manifest), ...auxiliary]
+          .sort((a, b) => (a.path < b.path ? -1 : 1))
```
Comment (FrederikBolding) [inline assets/sip-124/implementation/implementation.ts:19]: We may wanna include the failure case for identical paths in this reference implementation and the SIP

File: SIPS/sip-124.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
+---
+sip: 124
```
[the remark this answers is not in the document]
Comment (ritave) [reply SIPS/sip-124.md:2]: There would already be a collision between open PR for sip-18 (#122) and this one.

I'd prefer to add a comment to SIP-1 about how the number of the SIP correlates to PR number

File: assets/sip-124/implementation/implementation.ts (hunk 1 above)
Comment (ritave) [reply assets/sip-124/implementation/implementation.ts:6]: Not used in the example, I'll leave it out to simplify the example

File: SIPS/sip-124.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (hmalik88) [reply SIPS/sip-124.md:2]: I think I like the incrementing scheme better, if you're following the growth of the platform you have a better mental model of where a SIP is in relation to others. One just needs to be mindful of the other open SIP PRs when marking the SIP number.

File: SIPS/sip-124.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
+"OPTIONAL" written in uppercase in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)
+
+### Steps
+
+#### Checksum `snap.manifest.json`
+
+> This algorithm works on `snap.manifest.json` version `0.1`.
+
+<!-- TODO(ritave): Extract the algorithm from fast-json-stable-stringify and put it here -->
```
Comment (hmalik88) [inline SIPS/sip-124.md:29]: Feels weird to have a comment in the spec, can we either put the algorithm in here now or open an issue to add in later?

File: SIPS/sip-124.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+1. Parse `snap.manifest.json` into JavaScript object using `JSON.parse`.
+1. Delete `.result.source.shasum` field from `snap.manifest.json`.
```
Comment (hmalik88) [inline SIPS/sip-124.md:32]:
```suggestion
2. Delete `.result.source.shasum` field from `snap.manifest.json`.
```

File: SIPS/sip-124.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+1. Convert rest of the structure back to JSON using [`fast-json-stable-stringify@^2.1.0`](https://www.npmjs.com/package/fast-json-stable-stringify) algorithm.
```
Comment (hmalik88) [inline SIPS/sip-124.md:33]:
```suggestion
3. Convert rest of the structure back to JSON using [`fast-json-stable-stringify@^2.1.0`](https://www.npmjs.com/package/fast-json-stable-stringify) algorithm.
```

File: SIPS/sip-124.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+1. Checksum the resulting string using [auxiliary file](#checksum-auxiliary-files) algorithm.
```
Comment (hmalik88) [inline SIPS/sip-124.md:34]:
```suggestion
4. Checksum the resulting string using [auxiliary file](#checksum-auxiliary-files) algorithm.
```

**Pull request merged**